File: Hamming/python/HammingS.py
def calculate_parity_bits(m):
    r = 1
    while 2**r < m + r + 1:
        r += 1
    return r

def generate_hamming_code(binary_plot):
    #print("Data bits:", binary_plot)
    m = len(binary_plot)
    #print("Number of data bits to be encoded:", m)	
    r = calculate_parity_bits(m)
    #print("Number of parity bits to be added:", r)
    n = m + r
    #print("Total number of bits in Hamming code:", n)

    hamming_code = [-1] * n

    for i in range(r):
        hamming_code[2**i - 1] = 0

    #print("Hamming code in the beginning:", hamming_code)

    true_message_bits_pos = []
    parity_bits_pos = []

    for i in range(n):
        if hamming_code[i] == -1:
            hamming_code[i] = binary_plot.pop(0)
            true_message_bits_pos.append(i + 1)
        else:
            parity_bits_pos.append(i + 1)

    #print("Position of message bits:", true_message_bits_pos)
    #print("Hamming code:", hamming_code)
    tablita = {}
    for i in true_message_bits_pos:
        #sum_powers_of_2(i)
        tablita[i] = sum_powers_of_2(i)

    #print(tablita)
    #print("Position of parity bits:", parity_bits_pos)
    for e in parity_bits_pos:
        #print(f"Calculating parity bit for position {e}: {calculate_parity_bit(tablita, e)}")
        #print('parity bit: ', e)
        calculate_parity_bit(tablita, e, hamming_code)
    
    #print("Hamming code:", hamming_code)
    return ''.join(str(e) for e in hamming_code)

def calculate_parity_bit(tablita, bit, trama):
    # regla par, si hay un numero impar de 1's es 1, si hay un numero par de 1's es 0
    l = []
    for key in tablita:
        if bit in tablita[key]:
            l.append(key - 1)
    #print(l)
    parity_check = []
    for i in l:
        parity_check.append(trama[i])
    trama[bit - 1] = even_parity(parity_check)
    #print(f"Parity bit for position {bit} with parity {parity_check}: {trama[bit - 1]}")

def even_parity(bits):
    count = 0
    for i in bits:
        if i == '1':
            count += 1
    if count % 2 == 0:
        return 0
    else:
        return 1

def odd_parity(bits):    
    count = 0
    for i in bits:
        if i == '1':
            count += 1
    if count % 2 == 0:
        return 1
    else:
        return 0

def is_power_of_2(n):
    return n != 0 and (n & (n - 1)) == 0

def sum_powers_of_2(number):
    if is_power_of_2(number):
        print(f"{number} is already a power of 2.")
        return

    current_number = number
    powers = []

    while current_number > 0:
        power_of_2 = 1
        acum = 0
        while power_of_2 <= current_number:
            power_of_2 *= 2
            acum += 1

        power_of_2 //= 2
        current_number -= power_of_2
        powers.append(acum - 1)

    #print(powers)
    positions = []
    for i in range(len(powers)):
        positions.append(2**powers[i])

    #print(f"Positions to check: {positions}")
    return positions

def ASCII_toBinary(letter):
    ascii_value = ord(letter)
    #print(' - letter: ', letter)
    #print(' - ascii: ', ascii_value)
    b = bin(ascii_value)[2:]
    return b

import random

def Noise(hamming):
    # probabilidad de ruido de 1.5%
    i = 0   
    while i < len(hamming):
        p = random.uniform(0, 1)
        #print('p: ', p)
        if p > 0.985:
            #print("RUIDO A LA VERGA RUIDO")
            if hamming[i] == '1':
                hamming[i] = '0'
            else:
                hamming[i] = '1'
        i += 1

    hamming = ''.join(str(e) for e in hamming)
    return hamming



"""entrada = input('Ingrese su mensaje: ')
cadena = list(entrada)
#print('Mensaje original: ', cadena)
cabena_binaria = []
for i in cadena:
    cabena_binaria.append(ASCII_toBinary(i))

#print('Trama en binario: ', cabena_binaria)
trama = []
for i in cabena_binaria:
    temp = generate_hamming_code(list(i))
    #print('Trama en código de hamming: ', temp)
    trama.append(temp)

#print('Trama en código de hamming: ', trama)
trama_noise = []
for i in trama:
    trama_noise.append(Noise(list(i)))

print('Trama con ruido: ', trama_noise)
#for e in trama_noise:
    #print(e, type(e))

from HammingR import main as HammingR"""


def message():
    entrada = input('Ingrese su mensaje: ')
    cadena = list(entrada)
    cadena_binaria = []
    for i in cadena:
        cadena_binaria.append(ASCII_toBinary(i))

    trama = []
    for i in cadena_binaria:
        temp = generate_hamming_code(list(i))
        trama.append(temp)

    trama_noise = []
    for i in trama:
        trama_noise.append(Noise(list(i)))

    return trama_noise

File: Hamming/java/HammingS.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.Scanner;
//import java.util.Arrays;

public class HammingS {

    public static int calculateParityBits(int m) {
        int r = 1;
        while (Math.pow(2, r) < m + r + 1) {
            r++;
        }
        return r;
    }

    public static String generateHammingCode(String binaryPlot) {
        //System.out.println("Trama original: " + binaryPlot);
        int m = binaryPlot.length();
        int r = calculateParityBits(m);
        int n = m + r;

        char[] hammingCode = new char[n];

        for (int i = 0; i < r; i++) {
            hammingCode[(int) Math.pow(2, i) - 1] = '0';
        }

        List<Integer> trueMessageBitsPos = new ArrayList<>();
        List<Integer> parityBitsPos = new ArrayList<>();
        //System.out.println("Trama: " + Arrays.toString(hammingCode));
        for (int i = 0; i < n; i++) {
            if (hammingCode[i] == 0) {
                trueMessageBitsPos.add(i + 1);
                hammingCode[i] = binaryPlot.charAt(0);
                binaryPlot = binaryPlot.substring(1);
            } else {
                parityBitsPos.add(i + 1);
            }
        }
        //System.out.println("Trama: " + Arrays.toString(hammingCode));
        //System.out.println("Posiciones de bits de paridad: " + parityBitsPos);
        //System.out.println("Posiciones de bits de mensaje: " + trueMessageBitsPos);
        Map<Integer, List<Integer>> table = new HashMap<>();
        for (int i : trueMessageBitsPos) {
            table.put(i, sumPowersOf2(i));
        }
        //System.out.println("Tabla: " + table);

        for (int e : parityBitsPos) {
            //System.out.println("Calculando bit de paridad para la posición " + e);
            calculateParityBit(table, e, hammingCode);
        }
        //System.out.println("Trama: " + Arrays.toString(hammingCode));

        return new String(hammingCode);
    }

    public static void calculateParityBit(Map<Integer, List<Integer>> table, int bit, char[] trama) {
        List<Integer> l = new ArrayList<>();
        for (int key : table.keySet()) {
            if (table.get(key).contains(bit)) {
                l.add(key-1);
            }
        }
        //System.out.println("Bits a comprobar: " + l);
        List<Character> parityCheck = new ArrayList<>();
        for (int i : l) {
            parityCheck.add(trama[i]);
        }
        trama[bit - 1] = evenParity(parityCheck);
    }

    public static char evenParity(List<Character> bits) {
        int count = 0;
        for (char c : bits) {
            if (c == '1') {
                count++;
            }
        }
        return (count % 2 == 0) ? '0' : '1';
    }

    public static boolean isPowerOf2(int n) {
        return n != 0 && (n & (n - 1)) == 0;
    }

    public static List<Integer> sumPowersOf2(int number) {
        if (isPowerOf2(number)) {
            System.out.println(number + " is already a power of 2.");
            return null;
        }

        int currentNumber = number;
        List<Integer> powers = new ArrayList<>();

        while (currentNumber > 0) {
            int powerOf2 = 1;
            int acum = 0;
            while (powerOf2 <= currentNumber) {
                powerOf2 *= 2;
                acum++;
            }

            powerOf2 /= 2;
            currentNumber -= powerOf2;
            powers.add(acum - 1);
        }

        List<Integer> positions = new ArrayList<>();
        for (int i = 0; i < powers.size(); i++) {
            positions.add((int) Math.pow(2, powers.get(i)));
        }

        return positions;
    }


    public static String[] ASCII_to_Binary(String ascii) {
        String[] binary = new String[ascii.length()];
        for (int i = 0; i < ascii.length(); i++) {
            binary[i] = Integer.toBinaryString(ascii.charAt(i));
        }
        return binary;
    }

    public static String Noise(String b){
        StringBuilder sb = new StringBuilder();
        int seed = (int) System.currentTimeMillis();
        Random r = new Random(seed);
        for(char c : b.toCharArray()){
            double d = r.nextDouble();
            //System.out.println(" Probability_: "+d);
            if(d > 0.985){
                //System.out.println("RUIDO A LA VERGA RUIDO");
                sb.append(c == '0' ? '1' : '0');
            }else{
                sb.append(c);
            }
        }
        return sb.toString();
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Ingrese el mensaje: ");
        String dummy = scanner.nextLine();
        //String dummy = "Dark Souls 3";
        String[] binario = ASCII_to_Binary(dummy);
        //System.out.print("\bBinario: " + Arrays.toString(binario));
        //System.out.print("Ingrese la trama: ");
        //String trama = scanner.nextLine();
        scanner.close();
        //System.out.println("Trama original: " + trama);
        //String hammingCode = generateHammingCode(trama);
        //System.out.println("Trama en código de Hamming: " + hammingCode);
        String[] hammingCode = new String[binario.length];
        for (int i = 0; i < binario.length; i++) {
            hammingCode[i] = generateHammingCode(binario[i]);
        }
        //System.out.println("\nTrama sin ruido: " + Arrays.toString(hammingCode));
        String[] hammingCodeNoise = new String[hammingCode.length];
        for (int i = 0; i < hammingCode.length; i++) {
            hammingCodeNoise[i] = Noise(hammingCode[i]);
        }
        System.out.println("Trama con ruido: " + Arrays.toString(hammingCodeNoise));
    }
}
